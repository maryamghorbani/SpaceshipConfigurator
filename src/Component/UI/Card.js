import React from 'react';

//import styles
import classes from './Card.module.scss'

const Card = props => {
    return (<div className={`${classes.card} ${props.className} ${props.active&& 'card-active'}`} onClick={props.onClick}>
            {props.children}
        </div>)
};

export default Card;