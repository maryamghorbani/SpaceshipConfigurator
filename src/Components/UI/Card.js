import React from 'react';

//import styles
import classes from './Card.module.scss'

const Card = props => {
    return (<div className={`${classes.card} ${props.className} ${props.active&& classes.cardActive}`} onClick={props.onClick}>
            {props.children}
        </div>)
};

export default Card;