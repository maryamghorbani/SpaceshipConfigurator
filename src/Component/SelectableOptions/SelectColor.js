import React from "react";

//import components
import Card from "../UI/Crd";

//import styles
import classes from './Box.module.scss'

const SelectColor = () => {
    return (
        <div>
            <p className={classes.boxTitle}>Select color:</p>
            <div className={classes.boxWrapper}>
                <Card className={classes.box}>1</Card>
                <Card className={classes.box}>2</Card>
                <Card className={classes.box}>3</Card>
            </div>
        </div>
    )
};

export default SelectColor;