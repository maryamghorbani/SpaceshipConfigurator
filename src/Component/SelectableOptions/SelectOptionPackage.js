import React, {useState} from "react";
import {Col, Container, Row} from "react-bootstrap";

//import components
import Card from "../UI/Card";

//import styles
import classes from './Box.module.scss'
import 'bootstrap/dist/css/bootstrap.min.css'

const SelectOptionPackage = (props) => {
    const {items, onReceivedItem} = {...props}
    const [indexCurrentItem, setIndexCurrentItem] = useState()

    const onSelectedItem = index => {
        setIndexCurrentItem(index)
        onReceivedItem(items[index])
    }

    return (
        <Container>
            <Row xs={12} className={classes.boxTitle}>Select option package:</Row>
            <Row xs={12}>
                {items?.map((item, index) => {
                    return (
                        <Col xs={12} sm={4}>
                            <Card key={Math.random()} className={classes.box}
                                  onClick={e => onSelectedItem(index)} active={index === indexCurrentItem}>
                                <p>{item.title}</p>
                                <p>{item.price > 0 ? `+${item.price}â‚¬` : <p></p> }</p>
                                <div className={classes.subBox}>
                                    {item.features.map(item => {
                                        return <div>{item}</div>
                                    })}
                                </div>
                            </Card>
                        </Col>
                    )
                })}
            </Row>
        </Container>
    )
};

export default SelectOptionPackage;