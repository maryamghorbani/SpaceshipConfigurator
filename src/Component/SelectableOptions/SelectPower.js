import React from "react";

//import components
import Card from "../UI/Crd";

//import styles
import classes from './Box.module.scss'

const SelectPower = () => {
    return (
        <div>
            <p className={classes.boxTitle}>Select power:</p>
            <div className={classes.boxWrapper}>
                <Card className={classes.box}>1</Card>
                <Card className={classes.box}>2</Card>
                <Card className={classes.box}>3</Card>
            </div>
        </div>
    )
};

export default SelectPower;