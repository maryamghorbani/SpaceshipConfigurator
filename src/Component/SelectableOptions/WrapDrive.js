import React from "react";

//import components
import Card from "../UI/Card";

//import styles
import classes from './Box.module.scss'
import grid from "../UI/grid.module.scss";

const WrapDrive = () => {
    return (
        <div>
            <p className={classes.boxTitle}>Select drive:</p>
            <div className={`${classes.boxWrapper} ${grid.grid}`}>
                <Card className={`${classes.box} ${grid.gridCol4}`}>
                    <p>No</p>
                    <p>+0€</p>
                </Card>
                <Card className={`${classes.box} ${grid.gridCol4}`}>
                    <p>Yes</p>
                    <p>+1000€</p>
                </Card>
            </div>
        </div>
    )
};

export default WrapDrive;