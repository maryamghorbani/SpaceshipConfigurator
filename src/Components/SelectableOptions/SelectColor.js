import React, {useState} from "react";
import {Col, Container, Row} from "react-bootstrap";

//import components
import Card from "../UI/Card";

//import styles
import classes from './Box.module.scss'
import 'bootstrap/dist/css/bootstrap.min.css'

const SelectColor = props => {

    const {items, onReceivedItem} = {...props}
    const [indexCurrentItem, setIndexCurrentItem] = useState()

    const onSelectedItem = index => {
        setIndexCurrentItem(index)
        onReceivedItem(items[index])
    }

    return (
        <Container>
            <Row xs={12} className={classes.boxTitle}>Select color:</Row>
            <Row data-testid='colors' xs={12}>
                <div data-testid="colorBox" className={classes.colorBox} style={{backgroundColor: '#fff'}}/>
                {items?.map((item, index) => {
                    return (
                        <Col xs={12} sm={4}>
                            <Card key={item.title} className={classes.box} onClick={e => onSelectedItem(index)} active={index === indexCurrentItem}>
                                <div data-testid="colorBox" className={classes.colorBox} style={{backgroundColor: `${item.color}`}}/>
                                <p data-testid=''>+{item.price}â‚¬</p>
                                <p data-testid='colorTitle'>{item.title}</p>
                            </Card>
                        </Col>
                    )
                })}
            </Row>
        </Container>
    )
};

export default SelectColor;