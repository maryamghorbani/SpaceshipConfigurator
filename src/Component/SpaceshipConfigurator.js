import React, {useEffect, useState} from "react";

//import component
import Card from "./UI/Card";
import Receipt from "./Receipt";

//import styles
import classes from './SpaceshipConfigurator.module.scss'
import grid from './UI/grid.module.scss'
import SelectColor from "./SelectableOptions/SelectColor";
import SelectPower from "./SelectableOptions/SelectPower";
import WrapDrive from "./SelectableOptions/WrapDrive";
import SelectOptionPakage from "./SelectableOptions/SelectOptionPakage";

const SpaceshipConfigurator = props => {

    // const [finalColorPrice, setFinalColorPrice] = useState();
    const [colors, setColors] = useState([])
    const [receiptItems, setReceiptItems] = useState({
        "basePrice": {
            title: "Base Price",
            price: 1000,
            prefix: "",
        },
        "color": {
            title: "color",
            price: 0,
            prefix: "+",
        },
        "power": {
            title: "power",
            price: 0,
            prefix: "+",
        },
        "wrapDrive": {
            title: "wrap drive",
            price: 0,
            prefix: "+",
        },
        "optionPackage": {
            title: "option package",
            price: 0,
            prefix: "+",
        },
    })
    const [receiptTotalPrice, setReceiptTotalPrice] = useState(0)

    // fetch colors
    useEffect(() => {
        setColors([
            {
                color: "#ffffff",
                price: 0,
                title: "Snow",
            },
            {
                color: "#FF7A00",
                price: 100,
                title: "Volcano",
            },
            {
                color: "#6BE4FF",
                price: 100,
                title: "Sky",
            }
        ])
    }, [])

    //
    useEffect(() =>{
        setReceiptTotalPrice(Object.keys(receiptItems).reduce(function(prev, current) {
            return prev + +receiptItems[current].price
        }, 0))
    }, [receiptItems])

    const onReceivedColor = color => {
        setReceiptItems(prevState => ({
            ...prevState, "color": {
                title: "color",
                price: color.price,
                prefix: "+",
            }
        }))
    }

    return (
        <Card className={classes.Wrapper}>
            <div>
                <p className={classes.title}>Spaceship Configurator</p>
                <div className={grid.grid}>
                    <div className={grid.gridCol8}>
                        <SelectColor colors={colors} onReceivedColor={onReceivedColor}/>
                        <SelectPower/>
                        <WrapDrive/>
                        <SelectOptionPakage/>
                    </div>
                    <Receipt className={grid.gridCol4} items={receiptItems} totalPrice={receiptTotalPrice}/>
                </div>
            </div>
        </Card>
    )
};

export default SpaceshipConfigurator;