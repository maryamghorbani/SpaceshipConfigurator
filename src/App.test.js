import React from "react";
import {shallow} from 'enzyme';
import '@testing-library/jest-dom/extend-expect';
import { render, screen, cleanup } from '@testing-library/react';

//import components
import SpaceshipConfigurator from "./Components/SpaceshipConfigurator";
import SelectColor from "./Components/SelectableOptions/SelectColor";
import SelectPower from "./Components/SelectableOptions/SelectPower";
import WrapDrive from "./Components/SelectableOptions/WrapDrive";
import SelectOptionPackage from "./Components/SelectableOptions/SelectOptionPackage";
import Card from "./Components/UI/Card";

afterEach(cleanup) //clean up everything after each test to avoid memory leaks

describe('<SpaceshipConfigurator /> with no props', () => {
    const container = shallow(<SpaceshipConfigurator/>);
    it('should match the snapshot', () => {
        expect(container.html()).toMatchSnapshot();
    });

    it('should have an title', () => {
        expect(container.find({className: 'title'}).text()).toEqual('Spaceship Configurator');
    });

    describe('<SelectColor /> with props', () => {
        const initialColorProps = {
            color: "#ffffff",
            price: 0,
            title: "Snow",
        };
        const colorComponent = shallow(<SelectColor {...initialColorProps} />);
        it('should have an title', () => {
            expect(colorComponent.find({className: 'boxTitle'}).text()).toEqual('Select color:');
        });
        it('should have Card component', () => {
            expect(colorComponent.find(<Card/>)).toBeTruthy();
        });
        it('should have a div via initialColorProps.color background', () => {
            const { getByTestId } = render(<SelectColor />)
            expect(getByTestId("colorBox")).toHaveStyle('background-color: #fff');
        })
    });
    describe('<SelectPower /> with props', () => {
        const powerComponent = shallow(<SelectPower/>);
        it('should have an title', () => {
            expect(powerComponent.find({className: 'boxTitle'}).text()).toEqual('Select power:');
        });
        it('should have Card component', () => {
            expect(powerComponent.find(<Card/>)).toBeTruthy();
        });
    });
    describe('<WrapDrive /> with props', () => {
        const driveComponent = shallow(<WrapDrive/>);
        it('should have an title', () => {
            expect(driveComponent.find({className: 'boxTitle'}).text()).toEqual('Wrap drive:');
        });
        it('should have Card component', () => {
            expect(driveComponent.find(<Card/>)).toBeTruthy();
        });
    });
    describe('<SelectOptionPackage /> with props', () => {
        const packageComponent = shallow(<SelectOptionPackage/>);
        it('should have an title', () => {
            expect(packageComponent.find({className: 'boxTitle'}).text()).toEqual('Select option package:');
        });
        it('should have Card component', () => {
            expect(packageComponent.find(<Card/>)).toBeTruthy();
        });
    });
});