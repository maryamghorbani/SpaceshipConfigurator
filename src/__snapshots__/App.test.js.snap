// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SpaceshipConfigurator /> with no props should match the snapshot 1`] = `"<div class=\\"card Wrapper undefined\\"><div class=\\"container\\"><div class=\\"title row\\">Spaceship Configurator</div><div class=\\"row\\"><div class=\\"col-md-8 col-12\\"><div class=\\"container\\"><div class=\\"boxTitle row row-cols-12\\">Select color:</div><div data-testid=\\"colors\\" class=\\"row row-cols-12\\"><div data-testid=\\"colorBox\\" class=\\"colorBox\\" style=\\"background-color:#fff\\"></div></div></div><div class=\\"container\\"><div class=\\"boxTitle row row-cols-12\\">Select power:</div><div class=\\"row row-cols-12\\"></div></div><div class=\\"container\\"><div class=\\"boxTitle row row-cols-12\\">Wrap drive:</div><div class=\\"row row-cols-12\\"></div></div><div class=\\"container\\"><div class=\\"boxTitle row row-cols-12\\">Select option package:</div><div class=\\"row row-cols-12\\"></div></div></div><div class=\\"col-md-4 col-12\\"><div><div class=\\"card receipt undefined\\"><div class=\\"row\\"><p>Base Price:</p><p>1000€</p></div><div class=\\"row\\"><p>color:</p><p>+0€</p></div><div class=\\"row\\"><p>power:</p><p>+0€</p></div><div class=\\"row\\"><p>wrap drive:</p><p>+0€</p></div><div class=\\"row\\"><p>option package:</p><p>+0€</p></div><div class=\\"row total\\"><p>Total:</p><p>0€</p></div></div></div></div></div></div></div>"`;
